6X: Check peer listener failed in IC-PROXY mode

interma opened the pull request:
This PR backported https://github.com/greenplum-db/gpdb/pull/16438 to 6x

----

In IC-PROXY mode, user doesn't get a notification when peer listener bind/listen failed. Because the related code is in IC-PROXY process, it only records warning logs. So, the user's query just hangs here silently. This behavior may make the user very confused.

In the commit, introduced a SHM variable to check the failure and give notification in time. An output example:
```
$ nc -l 2001
...
$ PGOPTIONS="-c gp_interconnect_type=proxy" psql demo
demo=# select count(*) from test;
ERROR:  SetupInterconnect: We are in IC_PROXY mode, but IC-Proxy Listener failed, please check. (ic_tcp.c:1296)  (seg1 slice1 127.0.1.1:6001 pid=189355) (ic_tcp.c:1296)
```

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

interma commented:
Dev pipeline failed, I am going to check it.

interma commented:
After investigation: 
Compared with 7x, 6x missed several icproxy commits (e.g. reload logic).
So, I made some little changes to my test case to make it pass.

And we have already opened https://github.com/greenplum-db/gpdb/issues/14485 to trace it: will align 6x's icproxy code to 7x when necessary (open another PR).

interma commented:
will merge https://github.com/greenplum-db/gpdb/pull/16586 (when it has been pushed) in this PR later.

interma commented:
Thanks for the review.

It has merged https://github.com/greenplum-db/gpdb/pull/16586 and wait pipeline to green:
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-report-error_icproxy-not-ready_6x-rhel8

kmacoskey commented on src/backend/cdb/motion/ic_tcp.c:
```diff
@@ -1277,6 +1277,9 @@ SetupTCPInterconnect(EState *estate)
 #ifdef ENABLE_IC_PROXY
+	/* check if current Segment's ICProxy listener failed */
+	if (pg_atomic_read_u32(ic_proxy_peer_listener_failed) > 0)
+		elog(ERROR, "SetupInterconnect: We are in IC_PROXY mode, but IC-Proxy Listener failed, please check.");
```
Is this error message too generic? Are there multiple ways it could fail and would it be possible to give more information than "IC-Proxy Listener failed"

interma replied:
Thank you, Kris.
It can also provide "ic-proxy listen at a port failed" (why it failed) and "please check the log" (to find which port it listened).
Do you think they are enough to explain the failure?

kmacoskey replied:
Yes. Information about which port and why it failed I think helps enrich the error message more. Thank you

soumyadeep2007 replied:
See https://github.com/greenplum-db/gpdb/pull/16438#discussion_r1379209694

interma replied:
Thank you all, I am going to address up it.

interma replied:
@kmacoskey @soumyadeep2007 
I have refined the error message in the third commit, please help to take a look.
Thanks.

kmacoskey replied:
:+1: Good changes, thank you!

interma commented:
Hi @soumyadeep2007, I think you are going to review this PR.
I have rebased and squashed the previous multi-commits into two commits:
* 9dd739c, the corresponding PR of main is https://github.com/greenplum-db/gpdb/pull/16438
* c4e71bc, the corresponding PR of main is https://github.com/greenplum-db/gpdb/pull/16586

Now, it should be easier to review, thanks.

soumyadeep2007 reviewed (approved):
LGTM! Please forward-port the change to main as well.

interma commented:
> LGTM! Please forward-port the change to main as well.

Thank you Deep, I opened https://github.com/greenplum-db/gpdb/pull/16717, please help to review.
